Device const qualifier check and fix

tbursztyka opened the pull request:
It's been a while since that has been checked.

Commit:
drivers: Check and fix device const qualifier on ISR

Re-running the script that checks for the const qualifier missing on
struct device ISR's parameter.

The script also changes the parameter 'arg' to 'dev' when relevant.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
hum... looks like when I am adding a reviewer, it removes another one. Is there a limit in reviewers or my browser is just messing up with github's ui?

henrikbrixandersen commented:
> hum... looks like when I am adding a reviewer, it removes another one. Is there a limit in reviewers or my browser is just messing up with github's ui?

There's a limit of 15 reviewers. It's causing problems elsewhere too.

henrikbrixandersen commented on drivers/entropy/entropy_cc13xx_cc26xx.c:
```diff
@@ -131,7 +131,7 @@ static int entropy_cc13xx_cc26xx_get_entropy(const struct device *dev,
 }
 
-static void entropy_cc13xx_cc26xx_isr(const void *arg)
+static void entropy_cc13xx_cc26xx_isr(const struct device *arg)
```
```suggestion
static void entropy_cc13xx_cc26xx_isr(const struct device *dev)
```

henrikbrixandersen commented on drivers/entropy/entropy_nrf5.c:
```diff
@@ -204,7 +204,7 @@ static void rng_pool_init(struct rng_pool *rngp, uint16_t size, uint8_t threshol
 }
 
-static void isr(const void *arg)
+static void isr(const struct device *arg)
```
```suggestion
static void isr(const struct device *dev)
```

henrikbrixandersen commented on drivers/espi/host_subs_npcx.c:
```diff
@@ -223,7 +223,7 @@ static inline uint8_t host_shd_mem_wnd_size_sl(uint32_t size)
 /* Host KBC sub-device local functions */
 #if defined(CONFIG_ESPI_PERIPHERAL_8042_KBC)
-static void host_kbc_ibf_isr(void *arg)
+static void host_kbc_ibf_isr(const struct device *arg)
```
```suggestion
static void host_kbc_ibf_isr(const struct device *dev)
```

henrikbrixandersen commented on drivers/espi/host_subs_npcx.c:
```diff
@@ -438,7 +438,7 @@ static void host_shared_mem_region_init(void)
 				defined(CONFIG_ESPI_PERIPHERAL_EC_HOST_CMD)
 /* Host pm (host io) sub-module isr function for all channels such as ACPI. */
-static void host_pmch_ibf_isr(void *arg)
+static void host_pmch_ibf_isr(const struct device *arg)
```
```suggestion
static void host_pmch_ibf_isr(const struct device *dev)
```

henrikbrixandersen commented on drivers/espi/host_subs_npcx.c:
```diff
@@ -438,7 +438,7 @@ static void host_shared_mem_region_init(void)
-static void host_pmch_ibf_isr(void *arg)
+static void host_pmch_ibf_isr(const struct device *arg)
 {
 	ARG_UNUSED(arg);
```
```suggestion
	ARG_UNUSED(dev);
```

henrikbrixandersen commented on drivers/espi/host_subs_npcx.c:
```diff
@@ -249,7 +249,7 @@ static void host_kbc_ibf_isr(void *arg)
-static void host_kbc_obe_isr(void *arg)
+static void host_kbc_obe_isr(const struct device *arg)
 {
 	ARG_UNUSED(arg);
```
```suggestion
	ARG_UNUSED(dev);
```

henrikbrixandersen commented on drivers/gpio/gpio_lpc11u6x.c:
```diff
@@ -418,7 +418,7 @@ static uint32_t gpio_lpc11u6x_get_pending_int(const struct device *dev)
 }
 
-static void gpio_lpc11u6x_isr(const void *arg)
+static void gpio_lpc11u6x_isr(const struct device *arg)
```
```suggestion
static void gpio_lpc11u6x_isr(const struct device *dev)
```

henrikbrixandersen commented on drivers/gpio/gpio_lpc11u6x.c:
```diff
@@ -418,7 +418,7 @@ static uint32_t gpio_lpc11u6x_get_pending_int(const struct device *dev)
-static void gpio_lpc11u6x_isr(const void *arg)
+static void gpio_lpc11u6x_isr(const struct device *arg)
 {
 	struct gpio_lpc11u6x_shared *shared =
 		(struct gpio_lpc11u6x_shared *)arg;
```
This looks weird.

henrikbrixandersen commented on drivers/i2c/i2c_cc13xx_cc26xx.c:
```diff
@@ -278,7 +278,7 @@ static int i2c_cc13xx_cc26xx_configure(const struct device *dev,
 }
 
-static void i2c_cc13xx_cc26xx_isr(const void *arg)
+static void i2c_cc13xx_cc26xx_isr(const struct device *arg)
```
```suggestion
static void i2c_cc13xx_cc26xx_isr(const struct device *dev)
```

henrikbrixandersen commented on drivers/i2c/i2c_cc13xx_cc26xx.c:
```diff
@@ -278,7 +278,7 @@ static int i2c_cc13xx_cc26xx_configure(const struct device *dev,
-static void i2c_cc13xx_cc26xx_isr(const void *arg)
+static void i2c_cc13xx_cc26xx_isr(const struct device *arg)
 {
 	const uint32_t base = get_dev_config(arg)->base;
 	struct i2c_cc13xx_cc26xx_data *data = get_dev_data(arg);
```
```suggestion
	const uint32_t base = get_dev_config(dev)->base;
	struct i2c_cc13xx_cc26xx_data *data = get_dev_data(dev);
```

henrikbrixandersen commented on drivers/i2c/i2c_mchp_xec.c:
```diff
@@ -457,7 +457,7 @@ static int i2c_xec_transfer(const struct device *dev, struct i2c_msg *msgs,
 }
 
-static void i2c_xec_bus_isr(void *arg)
+static void i2c_xec_bus_isr(const struct device *arg)
 {
 #ifdef CONFIG_I2C_SLAVE
 	struct device *dev = (struct device *)arg;
```
This should be cleaned up, imho.

tbursztyka replied:
the script  could not match this case. I'll update it.

tbursztyka replied:
looks like cocci does not like the #ifdef. it should be part of the '...' statement, but no.

henrikbrixandersen reviewed (changes requested):
I started commenting all the places where the `arg` name is kept for the ISR function argument to replace it with `dev`, but gave up seeing this is all over the place.

I would like to see all those `const struct device *arg` be renamed to `const struct device *dev` as that will trigger clean-up of the remaining code in the affected ISR functions (where we now have a `const struct device *arg` being typecasted to `const struct device *dev` and other, similar constructs.

tbursztyka commented on drivers/i2c/i2c_ite_it8xxx2.c:
```diff
@@ -868,7 +868,7 @@ static int i2c_it8xxx2_transfer(const struct device *dev, struct i2c_msg *msgs,
 }
 
-static void i2c_it8xxx2_isr(void *arg)
+static void i2c_it8xxx2_isr(const struct device *arg)
```
That kind of change my script does not do :)
I could probably add another rule to change the param name

Commit:
drivers: constify all device instances

Run cocci script to constify device instances.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
tests: constify all device instances

Run cocci script to constify device instances.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
pm: constify all device instances

Run cocci script to constify device instances.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

gmarull commented on tests/subsys/pm/power_mgmt/src/main.c:
```diff
@@ -411,15 +411,15 @@ void test_busy(void)
 {
-	pm_device_state_lock((struct device *)device_a);
+	pm_device_state_lock((const struct device *)device_a);
 	zassert_true(pm_device_state_is_locked(device_a), NULL);
 
 	testing_device_lock = true;
 	enter_low_power = true;
 
 	k_sleep(SLEEP_TIMEOUT);
 
-	pm_device_state_unlock((struct device *)device_a);
+	pm_device_state_unlock((const struct device *)device_a);
```
`device_a` is already a `const struct device*`

Commit:
soc: Check and fix const qualifier on ISR

Re-running the script that checks for the const qualifier missing on
ISR's parameter.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

ruuddw reviewed (dismissed):
ok for the ARC change, thanks.

henrikbrixandersen commented on drivers/counter/counter_esp32.c:
```diff
@@ -238,7 +238,7 @@ static const struct counter_driver_api counter_api = {
 
 static void counter_esp32_isr(void *arg)
```
```suggestion
static void counter_esp32_isr(const struct device *dev)
```

tbursztyka replied:
This one could not be found by the script, as it does note use IRQ_CONNECT macro.

henrikbrixandersen commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -593,7 +593,7 @@ static int IRAM_ATTR i2c_esp32_transfer(const struct device *dev, struct i2c_msg
 
 static void IRAM_ATTR i2c_esp32_isr(void *arg)
```
```suggestion
static void IRAM_ATTR i2c_esp32_isr(const struct device *dev)
```

henrikbrixandersen commented on drivers/interrupt_controller/intc_gd32_exti.c:
```diff
@@ -82,7 +82,7 @@ static const uint8_t line2irq[NUM_EXTI_LINES] = {
 };
 
-__unused static void gd32_exti_isr(void *isr_data)
+__unused static void gd32_exti_isr(const void *isr_data)
```
Was this intended?

tbursztyka replied:
the script works by steps, and it first checks about const qualifier on isr used through IRQ_CONNECT, so yes it's fine.

tbursztyka commented:
Commit d2118eb1bcb has flaws, I wonder how this passed CI as it is blocking mine...

gmarull commented:
> Commit [d2118eb](https://github.com/zephyrproject-rtos/zephyr/commit/d2118eb1bcb3bf991c076cad95fc5f0176785c9e) has flaws, I wonder how this passed CI as it is blocking mine...

https://github.com/zephyrproject-rtos/zephyr/pull/41626

erwango commented:
> > Commit [d2118eb](https://github.com/zephyrproject-rtos/zephyr/commit/d2118eb1bcb3bf991c076cad95fc5f0176785c9e) has flaws, I wonder how this passed CI as it is blocking mine...
> 
> #41626

I guess you meant https://github.com/zephyrproject-rtos/zephyr/pull/41999

sylvioalves commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -591,9 +591,8 @@ static int IRAM_ATTR i2c_esp32_transfer(const struct device *dev, struct i2c_msg
 }
 
-static void IRAM_ATTR i2c_esp32_isr(void *arg)
+static void IRAM_ATTR i2c_esp32_isr(const struct device *dev)
```
Both changes in ESP32 (counter and I2C) will add build warnings with this change. `esp_intr_alloc` requires `void *arg` in its signature. Can this be removed from this PR?

tbursztyka replied:
sure

MaureenHelm commented on drivers/can/can_stm32fd.c:
```diff
@@ -152,9 +152,8 @@ static int can_stm32fd_set_timing(const struct device *dev,
 }
 
-static void can_stm32fd_line_0_isr(void *arg)
+void can_stm32fd_line_0_isr(const struct device *dev)
```
Did you mean to drop `static`?

tbursztyka replied:
good catch, no indeed. Will check, that's not supposed to happen obviously

MaureenHelm commented on drivers/can/can_stm32fd.c:
```diff
@@ -164,11 +163,9 @@ static void can_stm32fd_line_0_isr(void *arg)
 }
 
-static void can_stm32fd_line_1_isr(void *arg)
+void can_stm32fd_line_1_isr(const struct device *dev)
```
Did you mean to drop `static`?

nashif commented on arch/arc/core/arc_smp.c:
```diff
@@ -114,7 +114,7 @@ void z_arc_slave_start(int cpu_num)
 #ifdef CONFIG_SMP
 
-static void sched_ipi_handler(const void *unused)
+static void sched_ipi_handler(const struct device *unused)
```
not following how this ends up a struct device, it is unrelated to devices...

tbursztyka replied:
Looks like I need to improve the cocci rule: if the parameter does not get casted to struct device into the function, it should not match.

carlocaione commented on arch/arm64/core/smp.c:
```diff
@@ -166,7 +166,7 @@ static void broadcast_ipi(unsigned int ipi)
 }
 
-void sched_ipi_handler(const void *unused)
+void sched_ipi_handler(const struct device *unused)
```
same here & co.

carlocaione commented on drivers/timer/arm_arch_timer.c:
```diff
@@ -18,9 +18,9 @@
 static uint64_t last_cycle;
 
-static void arm_arch_timer_compare_isr(const void *arg)
+static void arm_arch_timer_compare_isr(const struct device *dev)
```
This is not a device.

carlocaione commented on drivers/counter/counter_native_posix.c:
```diff
@@ -22,9 +22,9 @@ static struct counter_alarm_cfg pending_alarm;
 static const struct device *device;
 
-static void counter_isr(const void *arg)
+static void counter_isr(const struct device *dev)
```
again. Why?

carlocaione commented on drivers/entropy/entropy_nrf5.c:
```diff
@@ -201,11 +201,11 @@ static void rng_pool_init(struct rng_pool *rngp, uint16_t size, uint8_t threshol
 }
 
-static void isr(const void *arg)
+static void isr(const struct device *dev)
```
same

carlocaione commented on drivers/entropy/entropy_stm32.c:
```diff
@@ -278,11 +278,11 @@ static void rng_pool_init(struct rng_pool *rngp, uint16_t size,
 }
 
-static void stm32_rng_isr(const void *arg)
+static void stm32_rng_isr(const struct device *dev)
```
same

carlocaione commented on tests/kernel/gen_isr_table/src/main.c:
```diff
@@ -100,14 +100,14 @@ ISR_DIRECT_DECLARE(isr2)
 #endif
 
-void isr3(const void *param)
+void isr3(const struct device *param)
```
For sure I'm missing something here. But why all the ISR parameters must necessarily be a `struct device`?

carlocaione replied:
For sure I'm missing something here. But why all the ISR parameters must necessarily be a `struct device`?

carlocaione commented on tests/kernel/gen_isr_table/src/main.c:
```diff
@@ -100,14 +100,14 @@ ISR_DIRECT_DECLARE(isr2)
 #endif
 
-void isr3(const void *param)
+void isr3(const struct device *param)
```
For sure I'm missing something here. But why all the ISR parameters must necessarily be a `struct device`?

mnkp replied:
The parameter passed to the ISR is defined in `IRQ_CONNECT` macro [Defining a regular ISR](https://docs.zephyrproject.org/latest/reference/kernel/other/interrupts.html#defining-a-regular-isr). Most of the time drivers pass `struct device` but in a few cases that you pointed out that doesn't seem to be a case so it indeed should be fixed.

tbursztyka replied:
and the script I run did not check that the actual parameter is being used an actual struct device then. Now it's fixed.

carlocaione reviewed (dismissed):
The obvious problems were addressed. I still do not really like the fact that we are specializing the ISR argument to the actual usage, but I think that is a matter of personal preference.

tbursztyka commented:
@nashif  please, re-review

henrikbrixandersen commented:
@tbursztyka Please rebase and resolve the merge conflicts.

gmarull reviewed (dismissed):
PM bits LGTM

mbolivar-nordic commented:
This needs another rebase already :(

nashif commented:
> This needs another rebase already :( 

it has been a while since the last push :)

nashif commented on drivers/i2c/i2c_dw.c:
```diff
@@ -197,9 +197,8 @@ static inline void i2c_dw_write_byte_non_blocking(const struct device *dev, uint
 #endif
 
-static void i2c_dw_isr(void *arg)
+static void i2c_dw_isr(const struct device *port)
```
dev?

tbursztyka replied:
"port" was kept as this is the name used already for the device for the code (see below: arg was casted to a struct device *port).
